File: src/main/scala/com/exasol/parquetio/reader/RowReadSupport.scala
package com.exasol.parquetio.reader

import com.exasol.parquetio.data.Row
import com.exasol.parquetio.data.GenericRow
import com.exasol.parquetio.reader.converter.ParquetRootConverter

import org.apache.hadoop.conf.Configuration
import org.apache.parquet.hadoop.api.ReadSupport
import org.apache.parquet.hadoop.api.ReadSupport.ReadContext
import org.apache.parquet.io.api.RecordMaterializer
import org.apache.parquet.schema.MessageType

/**
 * A concrete implementation of [[org.apache.parquet.hadoop.api.ReadSupport]]
 * that materializes Parquet records into internal [[com.exasol.parquetio.data.Row]]
 * structure.
 */
final class RowReadSupport extends ReadSupport[Row] {
  override def prepareForRead(
    conf: Configuration,
    metadata: java.util.Map[String, String],
    messageType: MessageType,
    readContext: ReadContext
  ): RecordMaterializer[Row] =
    new RowRecordMaterializer(messageType, readContext)

  override def init(
    conf: Configuration,
    metadata: java.util.Map[String, String],
    messageType: MessageType
  ): ReadSupport.ReadContext = {
    val projection = conf.get(ReadSupport.PARQUET_READ_SCHEMA)
    val requestedSchema = ReadSupport.getSchemaForRead(messageType, projection)
    new ReadSupport.ReadContext(requestedSchema)
  }

  private[this] final class RowRecordMaterializer(
    messageType: MessageType,
    readContext: ReadContext
  ) extends RecordMaterializer[Row] {
    val rootConverter = ParquetRootConverter(messageType)
    override def getRootConverter(): ParquetRootConverter = rootConverter
    override def skipCurrentRecord(): Unit = getRootConverter().start()
    override def getCurrentRecord(): Row = new GenericRow(getRootConverter().getResult())
  }
}

File: src/main/java/com/exasol/parquetio/reader/RowReadSupport.java
package com.exasol.parquetio.reader;

import java.util.Map;

import org.apache.hadoop.conf.Configuration;
import org.apache.parquet.hadoop.api.ReadSupport;
import org.apache.parquet.io.api.RecordMaterializer;
import org.apache.parquet.schema.MessageType;

import com.exasol.parquetio.data.GenericRow;
import com.exasol.parquetio.data.Row;
import com.exasol.parquetio.reader.converter.ParquetRootConverter;

/**
 * Materializes Parquet records into internal {@link Row} values.
 */
@SuppressWarnings("deprecation")
final class RowReadSupport extends ReadSupport<Row> {
    @Override
    public RecordMaterializer<Row> prepareForRead(final Configuration configuration,
            final Map<String, String> metadata, final MessageType messageType, final ReadContext readContext) {
        return new RowRecordMaterializer(messageType);
    }

    @Override
    public ReadContext init(final Configuration configuration, final Map<String, String> metadata,
            final MessageType messageType) {
        final String projection = configuration.get(ReadSupport.PARQUET_READ_SCHEMA);
        final MessageType requestedSchema = ReadSupport.getSchemaForRead(messageType, projection);
        return new ReadContext(requestedSchema);
    }

    private static final class RowRecordMaterializer extends RecordMaterializer<Row> {
        private final MessageType messageType;
        private final ParquetRootConverter rootConverter;

        private RowRecordMaterializer(final MessageType messageType) {
            this.messageType = messageType;
            this.rootConverter = new ParquetRootConverter(messageType);
        }

        @Override
        public ParquetRootConverter getRootConverter() {
            return this.rootConverter;
        }

        @Override
        public void skipCurrentRecord() {
            getRootConverter().start();
        }

        @Override
        public Row getCurrentRecord() {
            return new GenericRow(this.messageType, getRootConverter().getResult());
        }
    }
}
